source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -885,9 +915,24 @@ void Filter::onUpstreamEvent(Network::ConnectionEvent event) {
 
 void Filter::onUpstreamConnection() {
   connecting_ = false;
-  // Re-enable downstream reads now that the upstream connection is established
-  // so we have a place to send downstream data to.
-  read_callbacks_->connection().readDisable(false);
+
+  if (!receive_before_connect_) {
+    // Re-enable downstream reads now that the upstream connection is established
+    // so we have a place to send downstream data to.
+    read_callbacks_->connection().readDisable(false);
+  }
+  // If we have received any data before upstream connection is established, send it to
+  // the upstream connection.
+  else if (early_data_buffer_.length() > 0) {
+    ENVOY_CONN_LOG(debug, "TCP:onUpstreamEvent() Flushing early data buffer to upstream",
+                   read_callbacks_->connection());
+    getStreamInfo().getUpstreamBytesMeter()->addWireBytesSent(early_data_buffer_.length());
+    upstream_->encodeData(early_data_buffer_, early_data_end_stream_);
+    ASSERT(0 == early_data_buffer_.length());
+
+    // Read enable the connection now that early data has been flushed to the upstream.
+    read_callbacks_->connection().readDisable(false);
```

Comment:
No there is a case that receive_before_connect is true but there was no early data so in that case the downstream is never read disabled. So we should have explicit checks before read enabling. Simplified the flow here a bit now, let me know what you think.

Reply:
I also wanted to do that but the issue with that is that there might be some other filter who has set read disabled. As of today, the read disabled state is a counter and has no notion of who set it. So if we check read disabled state and then enable it, then it might happen that TCP_PROXY modifies the state set by another filter.

Hence it is best that TCP_PROXY relies on it own logic on when it has set read disabled and read enable based on that only.

Reply:
Ack.